backend/src/statistics/statistics.service.ts

```diff
@@ -0,0 +1,158 @@
+      include: {
+        question: {
+          select: { difficulty: true },
+        },
+      },
+    });
+    const map = {
+      [Difficulty.EASY]: 0,
+      [Difficulty.MEDIUM]: 0,
+      [Difficulty.HARD]: 0,
+    };
+    for (const session of data) {
+      const { difficulty } = session.question;
+      map[difficulty] = map[difficulty] + 1;
+    }
+    return map;
+  }
+
+  private async getDurationSummaryByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: {
+        users: { some: { id: userId } },
+        endTime: { not: null },
+        startTime: {
+          gte: subDays(new Date(), 30),
+        },
+      },
+      include: {
+        question: {
+          select: { difficulty: true },
+        },
+      },
+    });
+    return data
+      .filter((session) => !!session.endTime)
+      .map((session) => {
+        return {
+          difficulty: session.question.difficulty as Difficulty,
+          timetaken: differenceInSeconds(
+            session.endTime || new Date(), // Should be defined by selection.
+            session.startTime,
+          ),
+          date: session.startTime,
+        };
+      });
+  }
+
+  private async getPeerSummaryByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: { users: { some: { id: userId } }, endTime: { not: null } },
+      include: {
+        question: {
+          select: { title: true },
+        },
+        users: true,
+      },
+    });
+    return data.map((session) => {
+      return {
+        userName: session.users.find((user) => user.id !== userId)?.name || "",
+        questionTitle: session.question.title,
+        date: session.startTime,
+      };
+    });
+  }
+
+  private async getHeatmapDataByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: {
+        users: { some: { id: userId } },
+        endTime: { not: null },
+        startTime: {
+          gte: subYears(new Date(), 1),
+        },
+      },
+    });
+    return data.map((session) => {
+      return { date: session.startTime };
+    });
+  }
+
+  private async getNetworkDataByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: {
+        users: { some: { id: userId } },
+        endTime: { not: null },
+      },
+      include: {
+        question: {
+          include: {
+            topics: true,
+          },
+        },
+      },
+    });
+    const topics = new Map<number, string>();
+    const relations = new Map<number, number>();
+    for (const session of data) {
+      const sortedTopics = session.question.topics.sort((l, r) => l.id - r.id);
+      for (const topic of sortedTopics) {
+        topics.set(topic.id, topic.name);
+      }
+
+      for (let idx = 0; idx < sortedTopics.length; idx++) {
+        for (let oth = idx + 1; oth < sortedTopics.length; oth++) {
```

Comment:
Not quite sure what oth is. I suppose its "other"? Might want to make it more verbose

Reply:
In addition, instead of having the topics directed to one another in a straight line (which I assume what this logic does), can we have the node with the largest number of questions pointing outwards to the other nodes? (ie .sort by value instead of by id)

Reply:
Will fix the style and the existing logic, but having the edges always point out from the largest node will significantly increase the processing time. Will it be significantly noticeable on the frontend?